torchtune/data/_instruct_templates.py

```diff
@@ -0,0 +1,147 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Mapping, Optional
+
+
+class InstructTemplate(ABC):
+    """
+    Interface for instruction templates. Each template should include the template
+    prompt with placeholders for the data inputs.
+    """
+
+    template = ""
+
+    @classmethod
+    @abstractmethod
+    def format(
+        cls, sample: Mapping[str, Any], column_map: Optional[Dict[str, str]] = None
+    ) -> str:
+        """
+        Format the prompt template with the given arguments.
+
+        Args:
+            sample (Mapping[str, Any]): a single data sample with various fields
+            column_map (Optional[Dict[str, str]]): a mapping from the expected
+                placeholder names in the template to the column names in the sample.
+                If None, assume these are identical. Note: if the sample output is not named
+                as "output" in the dataset, you always need to map it to "output" in column_map.
+
+        Returns:
+            The formatted prompt
+        """
+        pass
+
+
+class AlpacaInstructTemplate(InstructTemplate):
```

Comment:
Sorry if I misunderstood earlier, but I thought we're converting this to follow a similar structure to the chat formats i.e. tokenize instruction, input and response separately?

Reply:
No this is kinda our middle ground for now. This keeps the convenience of string formatting as an option for instruct datasets then casts to messages on the dataset side. Tbh this is one of the kinks I still want to iron out here, we could go all the way on the `Message` format but I'm hedging a bit for now cause I know people like the string-formatting of prompts. But yeah this may change in the future